contracts/modules/UniswapManager.sol

```diff
@@ -0,0 +1,159 @@
+pragma solidity ^0.5.4;
+import "../wallet/BaseWallet.sol";
+import "./common/BaseModule.sol";
+import "./common/RelayerModule.sol";
+import "./common/OnlyOwnerModule.sol";
+import "../exchange/ERC20.sol";
+import "../utils/SafeMath.sol";
+import "../utils/SafeMath.sol";
+import "../storage/GuardianStorage.sol";
+
+interface UniswapFactory {
+    function getExchange(address _token) external view returns(address);
+}
+
+/**
+ * @title UniswapManager
+ * @dev Contract enabling the wallet owner to pool tokens to a Uniswap liquidity pool.
+ * @author Julien Niset - <julien@argent.im>
+ */
+contract UniswapManager is BaseModule, RelayerModule, OnlyOwnerModule {
+
+    bytes32 constant NAME = "UniswapManager";
+
+    using SafeMath for uint256;
+
+    // The Uniswap factory contract
+    UniswapFactory uniswap;
+    // The Guardian storage 
+    GuardianStorage public guardianStorage;
+
+    /**
+     * @dev Throws if the wallet is locked.
+     */
+    modifier onlyWhenUnlocked(BaseWallet _wallet) {
+        require(!guardianStorage.isLocked(_wallet), "TT: wallet must be unlocked");
+        _;
+    }
+
+    constructor(
+        ModuleRegistry _registry, 
+        GuardianStorage _guardianStorage, 
+        address _uniswap
+    ) 
+        BaseModule(_registry, NAME) 
+        public 
+    {
+        guardianStorage = GuardianStorage(_guardianStorage);
+        uniswap = UniswapFactory(_uniswap);
+    }
+ 
+    /**
+     * @dev Adds liquidity to a Uniswap ETH-ERC20 pair.
+     * @param _wallet The target wallet
+     * @param _poolToken The address of the ERC20 token of the pair.
+     * @param _ethAmount The amount of ETH available.
+     * @param _tokenAmount The amount of ERC20 token available.
+     * @return the number of liquidity shares minted. 
+     */
+    function addLiquidityToUniswap(
+        BaseWallet _wallet, 
+        address _poolToken, 
+        uint256 _ethAmount, 
+        uint256 _tokenAmount
```

Comment:
why don't we allow the user to only provide a total value `_totalAmountInEth` that they are willing to deposit into Uniswap, and let the contract take care of calculating how this `_totalAmountInEth` should be divided into an `_ethAmount` and a `_tokenAmount` instead of requiring the user to provide both an `_ethAmount` and a `_tokenAmount`?

Reply:
The idea is to have maximum flexibility so the user can provide any amount of ETH and token and the contract will take care of the rest. So `_ethAmount` or `_tokenAmount` can be zero (which is the scenario you describe) but it doesn't need to be.

Note that we may choose to only allow the user to provide only ETH or token at the client level but at least the contracts can handle all scenarios.

Reply:
One problem I have with this approach is that a swap is almost always required even if the user owns enough ETH and tokens to add liquidity without having to do any swap. 

Example:
Assume 1 ETH = 100 DAI. User owns 100 ETH and 10,000 DAI. User wants to put ~2ETH in Uniswap. Client calls `addLiquidityToUniswap` with `_poolToken=100 DAI` and `_ethAmount=1ETH`. There will most likely be some swap going on, even though the user could have just used his existing ETH and DAI balances to do the operation.

The only case where a swap is not required is (assuming we add the `if` wrapping iI mentioned in my other comments) if the client manages to perfectly calculate `_ethAmount` and `_tokenAmount` such that `ethSwap` or `tokenSwap` is zero. But that would make the client implementation complicated and not very robust to changes given all the fees, rounding up, etc.

Reply:
Another approach could be:
1. client just specifies how much ETH he wants to add to the pool
2. contract calculates how many tokens are required
3. if user's token balance is lower than amount of tokens required, convert some ether to get the required balance
4. perform addLiquidity

Reply:
As discussed you have a fair point. In the end it boils down to the the UX we want to offer to our users. Since that UX is not properly defined yet I'm in favor of an implementation that covers as many use cases as possible.

To keep the current implementation while addressing your comment I have added a boolean flag `preventSwap` to the `addLiquidityWithUniswap()` method that can prevent swaps from happening by finding the lowest in value between `_ethAmount` and `_tokenAmount` and add liquidity based on that reference.